json_ref_dict/ref_pointer.py

```diff
@@ -102,4 +142,16 @@ def resolve_uri(uri: URI) -> Union[List, Dict, None, bool, str, int, float]:
         raise DocumentParseError(
             f"Failed to load base document of {uri}."
         ) from exc
-    return RefPointer(uri).resolve(document)
+
+    pointer = RefPointer(uri)
+    remote_uri, value = pointer.resolve_with_uri(document)
+    return remote_uri if remote_uri else uri, value
+
+
+def resolve_uri(uri: URI) -> ResolvedValue:
```

Comment:
There is a breaking change hear around caching.  Consider the following usage:
```python
resolve_uri.cache_clear()
```

This will fail because of an AttributeError.  _However_, simply adding an `lru_cache` decorator isn't sufficient to solve, because `resolve_uri.cache_clear` won't clear the inner cache.

Reply:
I think this would look roughly as follows:

```python
@lru_cache(maxsize=None)
def resolve_uri(uri, include_remote_uri=False):
    # get the pointer and doc
    remote_uri, value = pointer.resolve_with_uri(document)
    if include_key:
        return remote_uri, value
    return value

@typing.overload
def resolve_uri(uri: URI) -> ResolvedValue:
    ...

@typing.overload
def resolve_uri(uri: URI, include_key: Literal[True]) -> UriValuePair:
    ...
```

Reply:
I think `Literal` isn't available given the Python 3.6 support, can either use `bool` as a workaround (will give wrong typing in an edge case), or add [typing-extensions](https://pypi.org/project/typing-extensions/) as a dependency.

Reply:
A good catch! I ended up renaming the new resolver-function a bit better (hopefully now more readable). I did not want to break the 3.6 compatibility, so I went with cache decorator customization approach. The include_remote_uri argument approach would have also introduced additional cache misses as there is no easy way to exclude the boolean argument from cache key generation => similar cache hack would have been needed.

I hope you can live with this approach!